File: worker-api/src/glicko2.ts
// Glicko-2 Rating System Implementation
// Based on: http://www.glicko.net/glicko/glicko2.pdf

import type { Env, Bot, MatchParticipant } from './types';

// Glicko-2 constants
const SCALE = 173.7178; // Rating scale conversion factor
const TAU = 0.5; // System constant (constrains volatility change)
const DEFAULT_RATING = 1500;
const DEFAULT_RD = 350;
const DEFAULT_VOLATILITY = 0.06;

export interface Glicko2Rating {
  mu: number; // Mean rating (Glicko-2 scale)
  phi: number; // Rating deviation (Glicko-2 scale)
  sigma: number; // Volatility
}

/**
 * Convert rating to Glicko-2 scale
 */
export function toGlicko2(rating: number, rd: number): Glicko2Rating {
  return {
    mu: (rating - DEFAULT_RATING) / SCALE,
    phi: rd / SCALE,
    sigma: DEFAULT_VOLATILITY,
  };
}

/**
 * Convert from Glicko-2 scale to original scale
 */
export function fromGlicko2(g2: Glicko2Rating): { rating: number; rd: number } {
  return {
    rating: g2.mu * SCALE + DEFAULT_RATING,
    rd: g2.phi * SCALE,
  };
}

/**
 * Compute g(phi) function
 */
function g(phi: number): number {
  return 1 / Math.sqrt(1 + (3 * phi * phi) / (Math.PI * Math.PI));
}

/**
 * Compute E(mu, mu_j, phi_j) function
 */
function E(mu: number, mu_j: number, phi_j: number): number {
  return 1 / (1 + Math.exp(-g(phi_j) * (mu - mu_j)));
}

/**
 * Compute new rating deviation (Step 5/6)
 */
function computeNewPhi(phi: number, v: number): number {
  const phiSquared = phi * phi;
  const vInverse = 1 / v;
  return 1 / Math.sqrt(1 / phiSquared + vInverse);
}

/**
 * Iterative algorithm to compute new volatility (Step 5.4)
 */
function computeNewVolatility(
  sigma: number,
  phi: number,
  v: number,
  delta: number,
  tau: number = TAU
): number {
  let a = Math.log(sigma * sigma);
  const epsilon = 0.000001;

  const f = (x: number): number => {
    const expX = Math.exp(x);
    const tmp = phi * phi + v + expX;
    return (
      (expX * (delta * delta - phi * phi - v - expX)) / (2 * tmp * tmp) -
      (x - a) / (tau * tau)
    );
  };

  // Set initial bounds
  let A = a;
  let B: number;
  if (delta * delta > phi * phi + v) {
    B = Math.log(delta * delta - phi * phi - v);
  } else {
    let k = 1;
    while (f(a - k * tau) < 0) {
      k++;
    }
    B = a - k * tau;
  }

  // Illinois algorithm
  let fA = f(A);
  let fB = f(B);

  while (Math.abs(B - A) > epsilon) {
    const C = A + ((A - B) * fA) / (fB - fA);
    const fC = f(C);

    if (fC * fB <= 0) {
      A = B;
      fA = fB;
    } else {
      fA = fA / 2;
    }

    B = C;
    fB = fC;
  }

  return Math.exp(A / 2);
}

/**
 * Calculate rating updates for a bot after a match
 * @param bot The bot whose rating to update
 * @param opponents Array of opponent ratings and game outcomes (1=win, 0.5=draw, 0=loss)
 * @returns New rating values
 */
export function updateRating(
  bot: Bot,
  opponents: Array<{
    rating: number;
    rd: number;
    score: number;
  }>
): { rating: number; rd: number; volatility: number } {
  if (opponents.length === 0) {
    // No games played - increase RD over time (rating decay)
    const phi = bot.rating_deviation / SCALE;
    const newPhi = Math.min(Math.sqrt(phi * phi + bot.rating_volatility * bot.rating_volatility), 350 / SCALE);
    return {
      rating: bot.rating,
      rd: newPhi * SCALE,
      volatility: bot.rating_volatility,
    };
  }

  // Convert to Glicko-2 scale
  const g2 = toGlicko2(bot.rating, bot.rating_deviation);
  g2.sigma = bot.rating_volatility;

  // Step 3: Compute v (variance of game outcomes)
  let vInverse = 0;
  for (const opp of opponents) {
    const oppG2 = toGlicko2(opp.rating, opp.rd);
    const gPhi = g(oppG2.phi);
    const eValue = E(g2.mu, oppG2.mu, oppG2.phi);
    vInverse += gPhi * gPhi * eValue * (1 - eValue);
  }
  const v = 1 / vInverse;

  // Step 4: Compute delta (rating improvement)
  let deltaSum = 0;
  for (const opp of opponents) {
    const oppG2 = toGlicko2(opp.rating, opp.rd);
    const gPhi = g(oppG2.phi);
    const eValue = E(g2.mu, oppG2.mu, oppG2.phi);
    deltaSum += gPhi * (opp.score - eValue);
  }
  const delta = v * deltaSum;

  // Step 5: Compute new volatility
  const newSigma = computeNewVolatility(g2.sigma, g2.phi, v, delta);

  // Step 6: Update phi
  const phiStar = Math.sqrt(g2.phi * g2.phi + newSigma * newSigma);

  // Step 7: Update phi and mu
  const newPhi = 1 / Math.sqrt(1 / (phiStar * phiStar) + 1 / v);
  const newMu = g2.mu + newPhi * newPhi * deltaSum;

  // Convert back
  const result = fromGlicko2({ mu: newMu, phi: newPhi, sigma: newSigma });

  return {
    rating: result.rating,
    rd: result.rd,
    volatility: newSigma,
  };
}

/**
 * Update ratings for all participants in a completed match
 */
export async function updateMatchRatings(
  env: Env,
  matchId: string,
  participants: MatchParticipant[],
  winnerId: string | null
): Promise<void> {
  // Get all bots involved
  const botIds = participants.map((p) => p.bot_id);
  const placeholders = botIds.map(() => '?').join(',');

  const bots = await env.DB.prepare(
    `SELECT * FROM bots WHERE id IN (${placeholders})`
  )
    .bind(...botIds)
    .all<Bot>();

  if (!bots.results || bots.results.length !== participants.length) {
    throw new Error('Could not find all participant bots');
  }

  const botMap = new Map(bots.results.map((b) => [b.id, b]));

  // Calculate new ratings for each participant
  const updates: Array<{
    botId: string;
    rating: number;
    rd: number;
    volatility: number;
    won: boolean;
  }> = [];

  for (const participant of participants) {
    const bot = botMap.get(participant.bot_id);
    if (!bot) continue;

    // Build opponent list
    const opponents = participants
      .filter((p) => p.bot_id !== participant.bot_id)
      .map((opp) => {
        const oppBot = botMap.get(opp.bot_id)!;
        // Score: 1 for win, 0.5 for draw (if no winner), 0 for loss
        let score = 0.5;
        if (winnerId === participant.bot_id) {
          score = 1;
        } else if (winnerId === opp.bot_id) {
          score = 0;
        }
        return {
          rating: oppBot.rating,
          rd: oppBot.rating_deviation,
          score,
        };
      });

    const newRating = updateRating(bot, opponents);
    const won = winnerId === participant.bot_id;

    updates.push({
      botId: participant.bot_id,
      rating: newRating.rating,
      rd: newRating.rd,
      volatility: newRating.volatility,
      won,
    });
  }

  // Apply updates in a batch
  const now = new Date().toISOString();

  for (const update of updates) {
    // Update bot rating
    await env.DB.prepare(
      `UPDATE bots SET
        rating = ?,
        rating_deviation = ?,
        rating_volatility = ?,
        matches_played = matches_played + 1,
        matches_won = matches_won + ?,
        updated_at = ?
      WHERE id = ?`
    )
      .bind(
        update.rating,
        update.rd,
        update.volatility,
        update.won ? 1 : 0,
        now,
        update.botId
      )
      .run();

    // Update participant with rating change
    await env.DB.prepare(
      `UPDATE match_participants SET
        rating_after = ?,
        rating_deviation_after = ?
      WHERE match_id = ? AND bot_id = ?`
    )
      .bind(update.rating, update.rd, matchId, update.botId)
      .run();

    // Record rating history
    await env.DB.prepare(
      `INSERT INTO rating_history (id, bot_id, match_id, rating_before, rating_after, rating_deviation, recorded_at)
       VALUES (?, ?, ?, ?, ?, ?, ?)`
    )
      .bind(
        crypto.randomUUID(),
        update.botId,
        matchId,
        botMap.get(update.botId)!.rating,
        update.rating,
        update.rd,
        now
      )
      .run();
  }
}

File: cmd/acb-worker/glicko2.go
// Glicko-2 Rating System Implementation for acb-worker
// Based on: http://www.glicko.net/glicko/glicko2.pdf
package main

import "math"

const (
	glicko2Scale     = 173.7178
	glicko2Tau       = 0.5
	glicko2DefaultMu = 1500.0
	glicko2DefaultRD = 350.0
	glicko2Epsilon   = 1e-6
)

// Glicko2Rating represents a Glicko-2 rating.
type Glicko2Rating struct {
	Mu    float64 `json:"mu"`
	Phi   float64 `json:"phi"`
	Sigma float64 `json:"sigma"`
}

// DisplayRating returns the conservative rating estimate (mu - 2*phi).
func (r Glicko2Rating) DisplayRating() float64 {
	return r.Mu - 2*r.Phi
}

// toScale converts from Glicko-1 (mu,phi) to Glicko-2 internal scale.
func toScale(mu, phi float64) (float64, float64) {
	return (mu - glicko2DefaultMu) / glicko2Scale, phi / glicko2Scale
}

// fromScale converts from Glicko-2 internal scale back to Glicko-1.
func fromScale(mu2, phi2 float64) (float64, float64) {
	return mu2*glicko2Scale + glicko2DefaultMu, phi2 * glicko2Scale
}

// gFunc computes g(phi) = 1 / sqrt(1 + 3*phi^2/pi^2)
func gFunc(phi float64) float64 {
	return 1.0 / math.Sqrt(1.0+3.0*phi*phi/(math.Pi*math.Pi))
}

// eFunc computes E(mu, mu_j, phi_j)
func eFunc(mu, muJ, phiJ float64) float64 {
	return 1.0 / (1.0 + math.Exp(-gFunc(phiJ)*(mu-muJ)))
}

// computeVolatility implements Step 5 of Glicko-2 (Illinois algorithm).
func computeVolatility(sigma, phi, v, delta float64) float64 {
	a := math.Log(sigma * sigma)
	tau := glicko2Tau

	f := func(x float64) float64 {
		expX := math.Exp(x)
		tmp := phi*phi + v + expX
		return (expX*(delta*delta-phi*phi-v-expX))/(2*tmp*tmp) - (x-a)/(tau*tau)
	}

	A := a
	var B float64
	if delta*delta > phi*phi+v {
		B = math.Log(delta*delta - phi*phi - v)
	} else {
		k := 1.0
		for f(a-k*tau) < 0 {
			k++
		}
		B = a - k*tau
	}

	fA := f(A)
	fB := f(B)

	for math.Abs(B-A) > glicko2Epsilon {
		C := A + (A-B)*fA/(fB-fA)
		fC := f(C)
		if fC*fB <= 0 {
			A = B
			fA = fB
		} else {
			fA = fA / 2
		}
		B = C
		fB = fC
	}

	return math.Exp(A / 2)
}

type opponent struct {
	mu2   float64 // Glicko-2 scale
	phi2  float64 // Glicko-2 scale
	score float64 // 1.0 = win, 0.5 = draw, 0.0 = loss
}

// updateSingleRating updates a single player's rating given their opponents.
func updateSingleRating(r Glicko2Rating, opps []opponent) Glicko2Rating {
	if len(opps) == 0 {
		// No opponents: increase RD (rating decay)
		phi2 := r.Phi / glicko2Scale
		newPhi2 := math.Min(math.Sqrt(phi2*phi2+r.Sigma*r.Sigma), glicko2DefaultRD/glicko2Scale)
		return Glicko2Rating{
			Mu:    r.Mu,
			Phi:   newPhi2 * glicko2Scale,
			Sigma: r.Sigma,
		}
	}

	mu2, phi2 := toScale(r.Mu, r.Phi)

	// Step 3: Compute v (variance)
	vInverse := 0.0
	for _, o := range opps {
		gPhi := gFunc(o.phi2)
		e := eFunc(mu2, o.mu2, o.phi2)
		vInverse += gPhi * gPhi * e * (1 - e)
	}
	v := 1.0 / vInverse

	// Step 4: Compute delta
	deltaSum := 0.0
	for _, o := range opps {
		gPhi := gFunc(o.phi2)
		e := eFunc(mu2, o.mu2, o.phi2)
		deltaSum += gPhi * (o.score - e)
	}
	delta := v * deltaSum

	// Step 5: Compute new volatility
	newSigma := computeVolatility(r.Sigma, phi2, v, delta)

	// Step 6: Update phi*
	phiStar := math.Sqrt(phi2*phi2 + newSigma*newSigma)

	// Step 7: Update phi and mu
	newPhi2 := 1.0 / math.Sqrt(1.0/(phiStar*phiStar)+1.0/v)
	newMu2 := mu2 + newPhi2*newPhi2*deltaSum

	newMu, newPhi := fromScale(newMu2, newPhi2)

	return Glicko2Rating{
		Mu:    newMu,
		Phi:   newPhi,
		Sigma: newSigma,
	}
}

// UpdateRatings computes new ratings for all participants in a multi-player match.
// Scores are used pairwise: for each pair (i, j), player i gets:
//   - 1.0 if scores[i] > scores[j]
//   - 0.5 if scores[i] == scores[j]
//   - 0.0 if scores[i] < scores[j]
func UpdateRatings(ratings []Glicko2Rating, scores []float64) []Glicko2Rating {
	n := len(ratings)
	if n < 2 {
		return ratings
	}

	result := make([]Glicko2Rating, n)

	for i := 0; i < n; i++ {
		var opps []opponent
		for j := 0; j < n; j++ {
			if i == j {
				continue
			}

			var s float64
			switch {
			case scores[i] > scores[j]:
				s = 1.0
			case scores[i] == scores[j]:
				s = 0.5
			default:
				s = 0.0
			}

			mu2, phi2 := toScale(ratings[j].Mu, ratings[j].Phi)
			opps = append(opps, opponent{mu2: mu2, phi2: phi2, score: s})
		}

		result[i] = updateSingleRating(ratings[i], opps)
	}

	return result
}

// ComputeRatingUpdates computes rating updates for match participants.
// botIDs, currentRatings, and scores must all have the same length.
func ComputeRatingUpdates(botIDs []string, currentRatings []Glicko2Rating, scores []float64) []RatingUpdate {
	if len(botIDs) != len(currentRatings) || len(botIDs) != len(scores) {
		return nil
	}

	newRatings := UpdateRatings(currentRatings, scores)
	updates := make([]RatingUpdate, len(botIDs))

	for i := range botIDs {
		updates[i] = RatingUpdate{
			BotID:                 botIDs[i],
			Mu:                    newRatings[i].Mu,
			Phi:                   newRatings[i].Phi,
			Sigma:                 newRatings[i].Sigma,
			DisplayRating:         newRatings[i].DisplayRating(),
			RatingMuBefore:        currentRatings[i].Mu,
			RatingPhiBefore:       currentRatings[i].Phi,
			RatingDeviationChange: newRatings[i].Phi - currentRatings[i].Phi,
		}
	}

	return updates
}
